keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,847 @@
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.criteria.succeededIndexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `succeededIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// SuccessPolicy specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// Specified it must be immutable and works only for the Indexed jobs.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	// +listType=atomic
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Criteria represents the list of alternative criteria for declaring the jobs 
+	// as successful before its completion. Once any of the criteria are met,
+	// the "SuccessCriteriaMet" condition is added, and the lingering pods are removed.
+	// The terminal state for such a job has the "Complete" condition.
+	// Additionally, these criteria are evaluated in order; Once the job meets one of the criteria,
+	// other criteria are ignored.
+	//
+	// +optional
+	Criteria []SuccessPolicyCriterion
+}
+
+// SuccessPolicyCriteria describes a criterion for declaring a Job as succeeded.
+// Each criterion must have at least one of the "succeededIndexes" or "succeededCount" specified.
+type SuccessPolicyCriterion struct {
+	// SuccessPolicyCriterion specifies the set of indexes 
+	// which need to be contained in the actual set of succeeded indexes for the job.
+	// The list of indexes must come within 0 to ".spec.completions" and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	SucceededIndexes *string
+
+	// SucceededCount specifies the minimal size of the actual set of the succeeded indexes
+	// for the job.
+	// When specified it needs to be positive integer.
+	//
+	// +optional
+	SucceededCount *int
```

Comment:
Somehow I assumed they are mutually exclusive, such as `onExitCodes` and `onPodConditions` in the pod failure policy. 

If they are not mutually exclusive we need to clarify how to interpret interpret `succeededIndexes` together with `SucceededCount`. 

I can see two interpretations
A. simple logical AND 
B. `succeededCount` defines the expected number of succeeded indexes, but only within `succeededIndexes`

Initially, I thought we go with (A.) but in that case it never makes sense to specify `succeededCount<|succeededIndexes|`, which is done is story 2 & 3.

(B.) is ok for me, but then I would suggest:
1. adjust the API comment, maybe:
```
// SucceededCount specifies the minimal required size of the actual set of the succeeded indexes
// for the job. When SucceededCount is used along with SucceededIndexes, then the check is 
// constrained only to the set of indexes specified by SucceededIndexes.
```
2. add validation rule, that `SucceededCount <= |SucceededIndexes|`. Otherwise it would never be satisfied.

Reply:
That's a good point.

> If they are not mutually exclusive we need to clarify how to interpret interpret succeededIndexes together with SucceededCount.
> 
> I can see two interpretations
> A. simple logical AND
> B. succeededCount defines the expected number of succeeded indexes, but only within succeededIndexes
> 
> Initially, I thought we go with (A.) but in that case it never makes sense to specify succeededCount<|succeededIndexes|, which is done is story 2 & 3.

You' right. If we specify both `succeededCount` and `succeededIndexes`, those parameters must be `succeededCount<|succeededIndexes|`. I will apply your suggestions to API comments and validations.

Reply:
Done.

Reply:
@mimowo @soltysh I applied @mimowo's suggestions. PTAL.